modules/ds18b20/ds18b20.c

```diff
@@ -0,0 +1,467 @@
+ * @param  config_register: value of configuration register.
+ * @return Resolution in bits: 9, 10, 11, or 12 
+ */
+uint8_t DS18B20_ConfigRegToResolution(uint8_t config_register) {
+    return ((config_register & DS18B20_CONFIG_REGISTER_R0_R1_BITMASK) 
+        >> DS18B20_CONFIG_REGISTER_RESERVED_BITS) + 9;
+}
+
+DS18B20_Status DS18B20_Start(OneWire_t* OneWire, uint8_t *ROM) {
+    if (!OneWire || !ROM) return DS18B20_USAGE_ERROR;
+    if (DS18B20_Is(ROM)==DS18B20_INVALID_DEVICE) return DS18B20_INVALID_DEVICE;
+    
+    /* Reset line */
+    if (OneWire_Reset(OneWire) == ONEWIRE_FAILURE) return DS18B20_FAILURE;
+    /* Select ROM number */
+    OneWire_SelectWithPointer(OneWire, ROM);
+    /* Start temperature conversion */
+    OneWire_WriteByte(OneWire, DS18B20_CMD_CONVERTTEMP);
+    
+    return DS18B20_SUCCESS;
+}
+
+DS18B20_Status DS18B20_StartAll(OneWire_t* OneWire) {
+    if (!OneWire) return DS18B20_USAGE_ERROR;
+    /* Reset pulse, abort if presence pulse is not good */
+    if (OneWire_Reset(OneWire) == ONEWIRE_FAILURE) return DS18B20_FAILURE;
+    /* Skip rom */
+    OneWire_WriteByte(OneWire, ONEWIRE_CMD_SKIPROM);
+    /* Start conversion on all connected devices */
+    OneWire_WriteByte(OneWire, DS18B20_CMD_CONVERTTEMP);
+    return DS18B20_SUCCESS;
+}
+
+DS18B20_Status DS18B20_Read(OneWire_t* OneWire, uint8_t *ROM, float *destination) {
+    if (!OneWire || !ROM || !destination) return DS18B20_USAGE_ERROR;
+
+    uint16_t temperature;
+    uint8_t resolution;
+    int8_t digit;
+    bool minus = false;
+    float decimal;
+    uint8_t i = 0;
+    uint8_t data[DS18B20_READ_DATA_SIZE];
+    uint8_t crc;
+    
+    /* Check if device is DS18B20 */
+    if (DS18B20_Is(ROM)==DS18B20_INVALID_DEVICE) {
+        return DS18B20_INVALID_DEVICE;
+    }
+    
+    /* Check if line is released, if it is, then conversion is complete */
+    uint8_t is_conversion_done = 0;
+    OneWire_ReadBit(OneWire,&is_conversion_done);
+    if (!is_conversion_done) {
+        /* Conversion is not finished yet */
+        return DS18B20_CONVERSION_IN_PROGRESS; 
+    }
+
+    /* Reset line */
+    if (OneWire_Reset(OneWire) == ONEWIRE_FAILURE) return DS18B20_FAILURE;
+    /* Select ROM number */
+    OneWire_SelectWithPointer(OneWire, ROM);
+    /* Read scratchpad command by onewire protocol */
+    OneWire_WriteByte(OneWire, ONEWIRE_CMD_RSCRATCHPAD);
+    
+    /* Get data */
+    for (i = 0; i < DS18B20_READ_DATA_SIZE; i++) {
+        /* Read byte by byte */
+        OneWire_ReadByte(OneWire, &data[i]);
+    }
+    
+    /* Calculate CRC */
+    // OneWire_CRC8(data, 8, &crc);
+    crc = OneWire_LookupCRC8(data,8);
+
+    /* Check if CRC is ok */
+    if (crc != data[DS18B20_READ_CRC_BYTE]) {
+        /* CRC invalid */
+        return DS18B20_FAILURE;
+    }
+    
+    /* First two bytes of scratchpad are temperature values */
+    temperature = data[DS18B20_DATA_LSB] | (data[DS18B20_DATA_MSB] << DS18B20_READ_CRC_BYTE);

```

Comment:
nit: I realize DS18B20_READ_CRC_BYTE is equal to 8.  But it seems weird reading this.  I think this is one case where a magic number (i.e. 8) would be ok.